patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch

```diff
@@ -17,23 +17,189 @@
          {
              if (!this.field_70170_p.field_72995_K && (this.field_70963_i == null || this.field_70963_i.size() > 0))
              {
-@@ -540,6 +541,7 @@
+@@ -255,6 +256,8 @@
+     {
+         super.func_70014_b(p_70014_1_);
+         p_70014_1_.func_74768_a("Profession", this.func_70946_n());
++        String professionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(this.func_70946_n()).getName();
++        p_70014_1_.func_74778_a("professionName", professionName);
+         p_70014_1_.func_74768_a("Riches", this.field_70956_bz);
+         p_70014_1_.func_74768_a("Career", this.field_175563_bv);
+         p_70014_1_.func_74768_a("CareerLevel", this.field_175562_bw);
+@@ -283,14 +286,34 @@
+     public void func_70037_a(NBTTagCompound p_70037_1_)
+     {
+         super.func_70037_a(p_70037_1_);
+-        this.func_70938_b(p_70037_1_.func_74762_e("Profession"));
++        int professionId = p_70037_1_.func_74762_e("Profession");
++        String professionName = p_70037_1_.func_74779_i("professionName");
```

Comment:
Vanilla returns and empty string if you request a string that is not in the tag, I could add a hasKey check incase that changes in the future.